src/d3d12/d3d12_resource_pool_texture.cpp

```diff
@@ -90,6 +111,109 @@ namespace wr
 		}
 	}
 
+	void D3D12TexturePool::Stage()
+	{
+		size_t unstaged_number = m_unstaged_textures.size();
+
+		if (unstaged_number > 0)
+		{
+			if (m_is_staging)
+			{
+				d3d12::WaitFor(m_staging_fence);
+				m_is_staging = false;
+			}
+
+			std::vector<d3d12::TextureResource*> unstaged_textures;
+
+			auto itr = m_unstaged_textures.begin();
+
+			for (itr; itr != m_unstaged_textures.end(); ++itr)
+			{
+				unstaged_textures.push_back(static_cast<d3d12::TextureResource*>(itr->second));
+			}
+
+			for (int i = 0; i < d3d12::settings::max_threads; ++i) {
+
+				d3d12::CommandList* cmdlist = m_command_lists[i];
+
+				m_command_recording_futures[i] = m_thread_pool->Enqueue([i](
+					d3d12::CommandList* cmdlist,
+					std::vector<d3d12::TextureResource*>& unstaged_textures)
+				{
+
+					d3d12::Begin(cmdlist, 0);
+
+					for (int j = 0; j<unstaged_textures.size(); ++j)
+					{
+						if (j%d3d12::settings::max_threads != i)
+						{
+							continue;
+						}
+
+						d3d12::TextureResource* texture = unstaged_textures[j];
+						
+						size_t row_pitch, slice_pitch;
+
+						DirectX::ComputePitch(static_cast<DXGI_FORMAT>(texture->m_format), texture->m_width, texture->m_height, row_pitch, slice_pitch);
+
+						D3D12_SUBRESOURCE_DATA subresourceData = {};
+						subresourceData.pData = texture->m_allocated_memory;
+						subresourceData.RowPitch = row_pitch;
+						subresourceData.SlicePitch = slice_pitch;
+
+						UpdateSubresources(cmdlist->m_native, texture->m_resource, texture->m_intermediate, 0, 0, 1, &subresourceData);
+
+						texture->m_is_staged = true;
+					}
+
+					d3d12::End(cmdlist);
+				},
+					cmdlist, unstaged_textures);
+
+			}
+
+			m_command_recording_futures[m_command_recording_futures.size() - 1] =
+				m_thread_pool->Enqueue([this](
```

Comment:
MipMapping shouldn't be threaded like this. Should also split across x amount of frames. Just remove this from the thread pool for now. GenerateMipmaps needs refactoring. After that we will add multithreading support to that.